File: RangeQuery/SegmentTree.py
class SegmentTree:
    def __init__(self, values):
        self.data = [0 for _ in values] + values
        self.n = len(values)
        
        for idx in reversed(range(1, self.n)):
            self.data[idx] = min(self.data[2*idx], self.data[2*idx+1])

    def update(self, idx, value):
        idx += self.n
        self.data[idx] = value

        while idx > 1:
            idx //= 2
            self.data[idx] = min(self.data[2*idx], self.data[2*idx+1])

    def minimum(self, left, right):
        left += self.n
        right += self.n
        minimum = self.data[left]

        while left < right:
            if left % 2:
                minimum = min(minimum, self.data[left])
                left += 1
            if right % 2:
                right -= 1
                minimum = min(minimum, self.data[right])
            left //= 2
            right //= 2

        return minimum


if __name__ == '__main__':
    st = SegmentTree([1, 5, 3, 7, 5])
    print(st.minimum(0, 3))
    print(st.minimum(1, 3))
    print(st.minimum(1, 2))

    st.update(0,3)
    st.update(2,4)

    print()
    print(st.minimum(0, 3))
    print(st.minimum(1, 3))
    print(st.minimum(1, 2))

File: RangeQuery/SegmentTree.java
import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

public class SegmentTree
{
    private ArrayList<Integer> data;
    private int n;

    public SegmentTree(List<Integer> arr) {
        n = arr.size();
        data = new ArrayList<Integer>(2 * n);

        for (int idx = 0; idx < n; idx++) {
            data.add(0);
        }
        for (int idx = 0; idx < n; idx++) {
            data.add(arr.get(idx));
        }
        for (int idx = n-1; idx > 0; idx--) {
            data.set(idx, Math.min(data.get(2*idx), data.get(2*idx+1)));
        }
    }

    public void update(int idx, int value) {
        idx += n;
        data.set(idx, value);
        
        while (idx > 1) {
            idx /= 2;
            data.set(idx, Math.min(data.get(2*idx), data.get(2*idx+1)));
        }
    }

    public int minimum(int left, int right) {
        left += n;
        right += n;
        int min = Integer.MAX_VALUE;

        while (left < right) {
            if ((left & 1) == 1) {
                min = Math.min(min, data.get(left));
                left++;
            }
            if ((right & 1) == 1) {
                right--;
                min = Math.min(min, data.get(right));
            }
            left >>= 1;
            right >>= 1;
        }
        
        return min;
    }

    public static void main(String[] args) {
        SegmentTree st = new SegmentTree(Arrays.asList(5, 2, 3, 1, 4));
        for (int i = 0; i < 5; i++) {
            System.out.println(st.minimum(i, i+1));
        }
        System.out.println(st.minimum(i, i+1));

        System.out.println(st.minimum(1, 4));
        st.update(3, 10);
        System.out.println(st.minimum(1, 4));
        System.out.println(st.minimum(0, 5));
        st.update(4, 0);
        System.out.println(st.minimum(1, 4));
        System.out.println(st.minimum(0, 5));
	}
}
